ingest/src/main/java/com/microsoft/azure/kusto/ingest/IngestionUtils.java

```diff
@@ -52,4 +55,16 @@ public static StreamSourceInfo resultSetToStream(ResultSetSourceInfo resultSetSo
         ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(byteArrayOutputStream.toByteArray());
         return new StreamSourceInfo(byteArrayInputStream, false, resultSetSourceInfo.getSourceId());
     }
+
+    public static byte[] readBytesFromInputStream(InputStream value, int testByteArraySize) throws IOException {
```

Comment:
The semantics for this method are wrong.
This will always read the entire buffer to memory. Unless I'm missing something, `testByteArraySize` doesn't do anything except setting the size of the buffer for each read.

The semantics we want is to read a maximum of `testByteArraySize` (or a better name) , or less if the stream ends before.

Look at the python implementation:
```python
def read_until_size_or_end(stream: IO[AnyStr], size: int) -> io.BytesIO:
    pos = 0
    result = io.BytesIO()
    while True:
        try:
            returned = stream.read(size - pos)
            pos += len(returned)
            result.write(returned)

            if len(returned) == 0 or pos == size:
                result.seek(0, io.SEEK_SET)
                return result

        except BlockingIOError:
            continue
```





Now the reason it didn't fail the test, is because the SequenceStream we put it in is opaque, and we can't know which bytes come from the byte array and which come from the original stream.
So I can't think of a good way to test it

Reply:
That's the openjdk implementation BTW
```java
    public static byte[] readNBytes(InputStream is, int len) throws IOException {
        if (len < 0) {
            throw new IllegalArgumentException("len < 0");
        }

        List<byte[]> bufs = null;
        byte[] result = null;
        int total = 0;
        int remaining = len;
        int n;
        do {
            byte[] buf = new byte[Math.min(remaining, DEFAULT_BUFFER_SIZE)];
            int nread = 0;

            // read to EOF which may read more or less than buffer size
            while ((n = is.read(buf, nread,
                    Math.min(buf.length - nread, remaining))) > 0) {
                nread += n;
                remaining -= n;
            }

            if (nread > 0) {
                if (MAX_BUFFER_SIZE - total < nread) {
                    throw new OutOfMemoryError("Required array size too large");
                }
                total += nread;
                if (result == null) {
                    result = buf;
                } else {
                    if (bufs == null) {
                        bufs = new ArrayList<>();
                        bufs.add(result);
                    }
                    bufs.add(buf);
                }
            }
            // if the last call to read returned -1 or the number of bytes
            // requested have been read then break
        } while (n >= 0 && remaining > 0);

        if (bufs == null) {
            if (result == null) {
                return new byte[0];
            }
            return result.length == total ?
                result : Arrays.copyOf(result, total);
        }

        result = new byte[total];
        int offset = 0;
        remaining = total;
        for (byte[] b : bufs) {
            int count = Math.min(b.length, remaining);
            System.arraycopy(b, 0, result, offset, count);
            offset += count;
            remaining -= count;
        }

        return result;
    }
```

Reply:
If your intent was to read exactly the specified number of bytes, why not use `java.io.InputStream.read(byte[] b, int off, int len)`?

Reply:
Nevermind, I see `readNBytes` calls that, but does more, presumably per the requirements you were describing.

Reply:
Specifically, read waits until any data is available and then returns whatever it got.
We need to keep reading until we're at the end of the stream or we passed the size.

Reply:
OK I updated it to the semantics you describe, instead of reading everything.